Automattic/woocommerce-payments
# Add 3DS support for pay for order page

This PR adds support for authentication payments using 3DS on "Pay for Order" pages. It uses almost exactly the same flow as introduced in the original 3DS PR, #606.

Fixes #686.

The difference between the checkout page and the "Pay for Order" page is that when WC Core redirects to the redirect URL after the payment was started, the redirect causes a page reload. This means that we need to check for the need for authentication when loading the page. This is done using the `maybeShowAuthenticationModal()` call.

The page reload also means that the credit card form will no longer have the credit card the user entered, so we need to hide and block the form while we wait for the user to authenticate. This is done using

```
blockUI( $( '#order_review' ) );
$( '#payment' ).hide( 500 );
```

I didn't follow the path laid out in the Stripe extension, introduced in stripe/pull/826. The main reason for the departure:

In Stripe, the way the credit card form is removed from being displayed is by removing Stripe from the list of available gateways. Removing WCPay from the list of available gateways in WCPay because the script `wcpay-checkout` is loaded using the `payment_fields()` function. This function is called by WC Core when WCPay is one of the available gateways, so if we remove WCPay from the list of gateways, our script doesn't load. This works in Stripe because the equivalent script is loaded in a different way.

Given this, I also figured we could do these things differently:
- redirect to the same URL and get the client secret using the hash again, instead of from hidden input
- we could remove the credit card form using JS instead of WP filters

In `update_order_status()` there's a new check to make sure that the intent saved to the order & the intent for the client secret used is the same. This is done in f203f04 by sending back the intent ID in the AJAX call to `update_order_status`.

# More Background Info
<details>
"Pay for Order" pages exist at URLs that look like this:

`example.com/checkout/order-pay/135/?pay_for_order=true&key=wc_order_123`

Payments for this are processed using `WC_Form_Handler::pay_action()`. The flow in WC Core is similar to process_payment, which is what's used when checking out.

In pay_action(), if the payment result is successful, the user is redirected to the page the payment gateway specifies:

```
if ( isset( $result['result'] ) && 'success' === $result['result'] ) {
	$result = apply_filters( 'woocommerce_payment_successful_result', $result, $order_id );

	wp_redirect( $result['redirect'] );
	exit;
}
```
_Found in https://github.com/woocommerce/woocommerce/blob/e9f71ac24cf0edcd2ac8b1539d1187e63b16fcc1/includes/class-wc-form-handler.php#L426-L431_

WC Core expects a redirect URL when a payment is successful. But, when a payment needs authentication it's not quite successful _yet_. For _checkout_ 3DS support, we have a workaround for this - we specify a hash instead of a new URL to redirect to. That helps in two ways:

- the user stays on the same page as before, making the checkout process look seamless (no page refresh during checkout)
- we are able to send information we need in the hash - the client secret needed to proceed with the authorization, and the order ID

However, the Pay for Order page has a GET parameter. This causes the page to reload, even if using only a hash as the redirect URL. This means that the credit card form gets re-initialized and the credit card number the user entered will no longer be displayed.

We can solve this by doing these things:

- Remove the credit card form from the page while we wait for the user to authenticate
- Block and hide the credit card form
</details>

# Testing instructions

1. Create an order either manually or by trying to check out and submitting a failing payment
2. Go to the Edit Order page for that order
3. Copy the link that says `Customer payment page →` 
4. That's the "Pay for Order" page link, go there to try out different payments
5. Try a 3DS payment using `4000002760003184` (auth required test cc) or `4000008260003178` (auth required, but fails)

# Issues and questions

- In Stripe I see we use `wc-credit-card-form-init` - I couldn't figure out why this was needed
